lib/services/unified_resource.go

```diff
@@ -214,7 +235,37 @@ func newWatcher(ctx context.Context, resourceCache *UnifiedResourceCache, cfg Re
 }
 
 func resourceKey(resource types.Resource) []byte {
-	return backend.Key(prefix, resource.GetName(), resource.GetKind())
+	var name, kind string
+	// set the kind to the appropriate "contained" type, rather than
+	// the container type. This better represents what resource is
+	// being sent to the ui
+	switch r := resource.(type) {
+	case types.AppServer:
+		app := r.GetApp()
+		if app != nil {
+			name = app.GetName()
+			kind = types.KindApp
+		}
+	case types.SAMLIdPServiceProvider:
+		name = r.GetName()
+		kind = types.KindApp
+	case types.KubeServer:
+		cluster := r.GetCluster()
+		if cluster != nil {
+			name = r.GetCluster().GetName()
+			kind = types.KindKubernetesCluster
+		}
+	case types.DatabaseServer:
+		db := r.GetDatabase()
+		if db != nil {
+			name = db.GetName()
+			kind = types.KindDatabase
+		}
+	default:
+		name = resource.GetName()
+		kind = resource.GetKind()
+	}
+	return backend.Key(prefix, name, kind)
```

Comment:
This is currently broke and I'm trying to fix it. When a resource is added, it's "contained resource" exists, but if something gets updated, like `AppServer`, then the event is sent without its contained resource which means I can't generate the resource key again. The frontend only receives the contained resource and uses it as the ID for pinning.

So, backend updates don't know about the 'contained' resource, and the frontend doesn't know about the 'container' resource. I have to figure out a way to connect them again.

You can ignore this PR until I figure it out

Reply:
Once https://github.com/gravitational/teleport/pull/32709 is in, I can refactor to use the `resourceSortKey.ByName` to fetch the correct resource by the pinned resource ID.

Reply:
Refactored using `resourceSortKey` and added the matcher.